### PR TITLE
Docker clang

### PR DESCRIPTION
**Pull request by rattrap**
CMAKE + LLVM + CLANG

### CODE REVIEW

**Review by MatusKysel — COMMENTED** [review]
I think something like `export CXX="clang++-10" CC="clang-10"` works better than the cmake

**Review by gaoqi8649 — COMMENTED** [review]
I see you add clang in Dockerfile. How to make sure Linux and Mac have same version in local?

File: Dockerfile [hunk 1]
```diff
@@ -63,7 +61,14 @@ ARG BUILD_OUTPUT_DIR
...
 RUN conan remote add -f bincrafters "https://api.bintray.com/conan/bincrafters/public-conan"
-RUN conan install -if $BUILD_OUTPUT_DIR --build missing -s build_type=Debug .
+RUN conan profile new clang --detect
+RUN conan profile update settings.compiler=clang clang
+RUN conan profile update settings.compiler.version=10 clang
+RUN conan profile update settings.compiler.libcxx=libstdc++11 clang
+RUN conan profile update env.CC=/usr/lib/llvm-10/bin/clang clang
+RUN conan profile update env.CXX=/usr/lib/llvm-10/bin/clang++ clang
+
+RUN conan install -if $BUILD_OUTPUT_DIR --build missing -s build_type=Debug -pr=clang .
```
Comment (sfrique) [inline Dockerfile:71]: Is there a reason for this not to be in a single `RUN` step?

**Comment by rattrap** [comment]
> I think something like `export CXX="clang++-10" CC="clang-10"` works better than the cmake

@MatusKysel done

File: Dockerfile (hunk 1 above)
Comment (rattrap) [reply Dockerfile]: No. Changed

**Comment by rattrap** [comment]
> I see you add clang in Dockerfile. How to make sure Linux and Mac have same version in local?

@gaoqi8649 I guess we can find a way to do that but this wasn't the purpose of this pull request and having clang-format and clang-tidy be downloaded each time wasn't the way to go.

File: conanfile.py [hunk 2]
```diff
@@ -53,11 +24,7 @@ def requirements(self):
         self.requires("snappy/1.1.8")
         self.requires("zstd/1.4.4")
         self.requires("lz4/1.9.2")
-        self.requires("libjson-rpc-cpp/1.1.0@bincrafters/stable")
-
-        # if it darwin we will check for some clang utils
-        if platform.system() == "Darwin":
-            self._add_clang_utils_on_darwin()
+        self.requires("libjson-rpc-cpp/1.1.1@bincrafters/stable")
```
[the remark this answers is not in the document]
Comment (rattrap) [reply conanfile.py:27]: Was getting the following error for the old version:

```
#20 58.86 [ 75%] Building CXX object src/lib_json/CMakeFiles/jsoncpp_lib_static.dir/json_value.cpp.o
#20 58.86 [ 75%] Building CXX object src/lib_json/CMakeFiles/jsoncpp_lib_static.dir/json_reader.cpp.o
#20 58.86 [ 75%] Building CXX object src/lib_json/CMakeFiles/jsoncpp_lib_static.dir/json_writer.cpp.o
#20 59.51 /root/.conan/data/jsoncpp/1.8.4/theirix/stable/build/00da4c182589defdef487951eb99550c66290ac0/source_subfolder/src/lib_json/json_value.cpp:73:15: error: implicit conversion from 'long' to 'double' may lose precision [-Werror,-Wimplicit-int-float-conversion]
#20 59.51   return d >= min && d <= max;
#20 59.51            ~~ ^~~
#20 59.51 /root/.conan/data/jsoncpp/1.8.4/theirix/stable/build/00da4c182589defdef487951eb99550c66290ac0/source_subfolder/src/lib_json/json_value.cpp:786:25: note: in instantiation of function template specialization 'Json::InRange<long, long>' requested here
#20 59.51     JSON_ASSERT_MESSAGE(InRange(value_.real_, minInt64, maxInt64),
#20 59.51                         ^
#20 59.51 /root/.conan/data/jsoncpp/1.8.4/theirix/stable/build/00da4c182589defdef487951eb99550c66290ac0/source_subfolder/src/lib_json/json_value.cpp:73:27: error: implicit conversion from 'long' to 'double' may lose precision [-Werror,-Wimplicit-int-float-conversion]
#20 59.51   return d >= min && d <= max;
#20 59.51                        ~~ ^~~
#20 59.51 /root/.conan/data/jsoncpp/1.8.4/theirix/stable/build/00da4c182589defdef487951eb99550c66290ac0/source_subfolder/src/lib_json/json_value.cpp:73:27: error: implicit conversion from 'unsigned long' to 'double' may lose precision [-Werror,-Wimplicit-int-float-conversion]
#20 59.51   return d >= min && d <= max;
#20 59.51                        ~~ ^~~
#20 59.51 /root/.conan/data/jsoncpp/1.8.4/theirix/stable/build/00da4c182589defdef487951eb99550c66290ac0/source_subfolder/src/lib_json/json_value.cpp:807:25: note: in instantiation of function template specialization 'Json::InRange<int, unsigned long>' requested here
#20 59.51     JSON_ASSERT_MESSAGE(InRange(value_.real_, 0, maxUInt64),
#20 59.51                         ^
#20 59.51 3 errors generated.
#20 59.51 make[2]: *** [src/lib_json/CMakeFiles/jsoncpp_lib_static.dir/build.make:76: src/lib_json/CMakeFiles/jsoncpp_lib_static.dir/json_value.cpp.o] Error 1
#20 59.51 make[2]: *** Waiting for unfinished jobs....
#20 59.88 make[1]: *** [CMakeFiles/Makefile2:133: src/lib_json/CMakeFiles/jsoncpp_lib_static.dir/all] Error 2
#20 59.88 make: *** [Makefile:141: all] Error 2
#20 59.89 jsoncpp/1.8.4@theirix/stable:
#20 59.89 jsoncpp/1.8.4@theirix/stable: ERROR: Package '00da4c182589defdef487951eb99550c66290ac0' build failed
#20 59.89 jsoncpp/1.8.4@theirix/stable: WARN: Build folder /root/.conan/data/jsoncpp/1.8.4/theirix/stable/build/00da4c182589defdef487951eb99550c66290ac0
#20 59.89 ERROR: jsoncpp/1.8.4@theirix/stable: Error in build() method, line 61
#20 59.89 	cmake.build()
#20 59.89 	ConanException: Error 2 while executing cmake --build '/root/.conan/data/jsoncpp/1.8.4/theirix/stable/build/00da4c182589defdef487951eb99550c66290ac0' '--' '-j8'
------
executor failed running [/bin/sh -c conan install -if $BUILD_OUTPUT_DIR --build missing -s build_type=Debug -pr=clang .]: exit code: 1
```

@kstdl figured out that it's fixed in the new one.

[the remark this answers is not in the document]
Comment (MatusKysel) [reply conanfile.py:27]: and then why not the latest available any issues there?

[the remark this answers is not in the document]
Comment (rattrap) [reply conanfile.py:27]: I will try building it with `1.3.0`

[the remark this answers is not in the document]
Comment (rattrap) [reply conanfile.py]: Updated

**Review by MatusKysel — APPROVED** [review]

**Review by sfrique — APPROVED** [review]

File: conanfile.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kstdl) [reply conanfile.py]: I had an linking issues with 1.2.0 and 1.3.0 versions earlier, so 1.1.0 was used originally

[the remark this answers is not in the document]
Comment (kstdl) [reply conanfile.py]: Strange, but it seems fine now with 1.3.0 version. So it is just OK

**Pull request merged**